src/libraries/Native/Unix/System.Globalization.Native/pal_collation.c

```diff
@@ -413,6 +423,110 @@ static const UCollator* GetCollatorFromSortHandle(SortHandle* pSortHandle, int32
     }
 }
 
+// If succeeded, U_SUCCESS(*pErr) should be true.
+// return -1 if couldn't borrow search handle from the SortHandle cache, otherwise, it return the slot number of the cache.
+static const int32_t GetSearchIteratorFromSortHandleUsingCollator(
+                        SortHandle* pSortHandle,
+                        const UCollator* pColl,
+                        const UChar* lpTarget,
+                        int32_t cwTargetLength,
+                        const UChar* lpSource,
+                        int32_t cwSourceLength,
+                        int32_t options,
+                        UErrorCode* pErr,
+                        UStringSearch** pSearchIterator)
+{
+    options &= CompareOptionsMask;
+    *pSearchIterator = pSortHandle->searchIteratorPerOption[options];
+
+    if (*pSearchIterator == NULL)
+    {
+        *pSearchIterator = usearch_openFromCollator(lpTarget, cwTargetLength, lpSource, cwSourceLength, pColl, NULL, pErr);
+        if (!U_SUCCESS(*pErr))
+        {
+            assert(FALSE && "Couldn't open the search iterator.");
+            return -1;
+        }
+
+        UStringSearch* pNull = NULL;
+        if (!pal_atomic_cas_ptr((void* volatile*)&pSortHandle->searchIteratorPerOption[options], *pSearchIterator, pNull))
+        {
+            usearch_close(*pSearchIterator);
+            *pSearchIterator = pSortHandle->searchIteratorPerOption[options];
+        }
+    }
+
+    assert(*pSearchIterator != NULL && "pSearch not expected to be null here.");
+
+    if (*pSearchIterator == USED_STRING_SEARCH || !pal_atomic_cas_ptr((void* volatile*)&pSortHandle->searchIteratorPerOption[options], USED_STRING_SEARCH, *pSearchIterator))
+    {
+        // The handle is in use. Create a temporary new one.
+        *pSearchIterator = usearch_openFromCollator(lpTarget, cwTargetLength, lpSource, cwSourceLength, pColl, NULL, pErr);
```

Comment:
We cache one sort handle per culture in managed code, right?  What does that mean for your measurements then, if only the first thread gets to use that cached sort handle's single cached search iterator?

Reply:
That is what I was saying the description above:

```
The search handles are a little bit different as it needs to include the source string we need to search in and the pattern string we need to search for. This require the search handle should be used exclusively and not be shared between 2 operations in same time.
```

The worst case when having a raise condition, we'll just create a new search handle as we already doing today.

Reply:
Should we have tests to explicitly try to trigger that race condition? I guess test runs will inevitably test it by accident, but if there is a bug there it will be less obvious because it will look like random failures.

Reply:
I have manually tested the race condition by running the search operation a million times simultaneously on number of threads equals to the number of procs in my machine and everything worked fine. I hesitated to add the test to our test suites as my machine was not usable when running the test.

Reply:
> The worst case when having a raise condition, we'll just create a new search handle as we already doing today.

That's my point. It doesn't seem so much like a race condition to me. If a server is doing a lot of concurrent during processing, it could be that just one core gets these speedups and the rest don't, so the benchmarks aren't necessarily meaningful for a solution at scale. Have I misunderstood? That's why I was asking about an alternative that is more consistent.

Reply:
fair enough. I'll think about it more to look if I can avoid this situation. The first I can think about is instead of caching one instance, I can cache more instance when I cannot get a free one.

Reply:
> by running the search operation a million times simultaneously

I guess I was thinking it could be pretty easy, if you start two threads that both do a large string search and are both released on an event. It will presumably hit the "not cached" case most of the time.

Reply:
I am currently changing the code to address @stephentoub feedback and will look adding a test as @danmosemsft suggesting too. Thanks for all feedback I am getting here.